src/vmm/src/devices/virtio/vsock/packet.rs

```diff
@@ -77,40 +77,130 @@ pub struct VsockPacketHeader {
+        self.len = len.to_le();
+        self
+    }
+
+    pub fn type_(&self) -> u16 {
+        u16::from_le(self.type_)
+    }
+
+    pub fn set_type(&mut self, type_: u16) -> &mut Self {
+        self.type_ = type_.to_le();
+        self
+    }
+
+    pub fn op(&self) -> u16 {
+        u16::from_le(self.op)
+    }
+
+    pub fn set_op(&mut self, op: u16) -> &mut Self {
+        self.op = op.to_le();
+        self
+    }
+
+    pub fn flags(&self) -> u32 {
+        u32::from_le(self.flags)
+    }
+
+    pub fn set_flags(&mut self, flags: u32) -> &mut Self {
+        self.flags = flags.to_le();
+        self
+    }
+
+    pub fn set_flag(&mut self, flag: u32) -> &mut Self {
+        self.set_flags(self.flags() | flag);
+        self
+    }
+
+    pub fn buf_alloc(&self) -> u32 {
+        u32::from_le(self.buf_alloc)
+    }
+
+    pub fn set_buf_alloc(&mut self, buf_alloc: u32) -> &mut Self {
+        self.buf_alloc = buf_alloc.to_le();
+        self
+    }
+
+    pub fn fwd_cnt(&self) -> u32 {
+        u32::from_le(self.fwd_cnt)
+    }
+
+    pub fn set_fwd_cnt(&mut self, fwd_cnt: u32) -> &mut Self {
+        self.fwd_cnt = fwd_cnt.to_le();
+        self
+    }
+}
+
 /// The vsock packet header struct size (the struct is packed).
 pub const VSOCK_PKT_HDR_SIZE: u32 = 44;
 
 // SAFETY: `VsockPacketHeader` is a POD and contains no padding.
 unsafe impl ByteValued for VsockPacketHeader {}
 
-/// Enum representing either a TX (e.g. read-only) or RX (e.g. write-only) buffer
-///
-/// Read and write permissions are statically enforced by using the correct `IoVecBuffer[Mut]`
-/// abstraction
-#[derive(Debug)]
-pub enum VsockPacketBuffer {
-    /// Buffer holds a read-only guest-to-host (TX) packet
-    Tx(IoVecBuffer),
-    /// Buffer holds a write-only host-to-guest (RX) packet
-    Rx(IoVecBufferMut),
-}
-
-/// Struct describing a single vsock packet.
-///
-/// Encapsulates the virtio descriptor chain containing the packet through the `IoVecBuffer[Mut]`
-/// abstractions.
-#[derive(Debug)]
-pub struct VsockPacket {
+// /// Struct describing a single vsock packet.
+// ///
+// /// Encapsulates the virtio descriptor chain containing the packet through the `IoVecBuffer[Mut]`
+// /// abstractions.
+#[derive(Debug, Default)]
+pub struct VsockPacketTx {
     /// A copy of the vsock packet's 44-byte header, held in hypervisor memory
     /// to minimize the number of accesses to guest memory. Can be written back
     /// to geust memory using [`VsockPacket::commit_hdr`] (only for RX buffers).
-    hdr: VsockPacketHeader,
+    pub hdr: VsockPacketHeader,
     /// The raw buffer, as it is contained in guest memory (containing both
     /// header and payload)
-    buffer: VsockPacketBuffer,
+    buffer: IoVecBuffer,
 }
 
-impl VsockPacket {
+// This is needed because IoVecBuffer contain *mut c_void type.
+// SAFETY:
+// Safe as only vmm thread will create and use this type.
+unsafe impl Send for VsockPacketTx {}
```

Comment:
this one doesn't need the impl, because `IoVecBuffer` is already `Send`

Reply:
done